### DISCUSSION TITLE
Class Inheritance with Foreign Keys in Sub-Classes (2.0 Annotated)

### DISCUSSION THREAD

**Discussion by daemonkeeper** [sqlalchemy/sqlalchemy#8959] [Usage Questions]
I've been experimenting (again) with sqlalchemy2; Today, I tried to declare a single-table class inherited model which would use the same columns in _some_ but not _all_ sub-classes for use as a foreign key into another table. 

The idea is to have a base model with sub-types that add specialized "mixin features" that, on their own, have foreign key relations. While that works fine if only one sub-class has the relations, there does not appear to be a way to have multiple (but not all) sub-classes using foreign key mixins. 

I do have a set-up where various sub-classes need various permutations of relationships into other tables. I've come up with this example to illustrate what I'm trying to achieve:

```python
from sqlalchemy import ForeignKey, create_engine
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Mapped, mapped_column, relationship, declared_attr, sessionmaker
from typing import Union

class BaseModel(MappedAsDataclass, DeclarativeBase, kw_only=True):
    ...

class TeamMixIn:
    team_id: Mapped[int|None] = mapped_column(ForeignKey("teams.id"), init=False, use_existing_column=True, repr=False)

    @declared_attr.cascading
    def team(_cls) -> Mapped["Team"]:
        return relationship(back_populates="user", init=False, repr=True)


class User(BaseModel):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user_type: Mapped[str] = mapped_column(init=False, repr=False)

    __mapper_args__ = {
        "polymorphic_identity": "user",
        "polymorphic_on": "user_type",
    }

class Engineer(User, TeamMixIn):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

class Accounting(User, TeamMixIn):
    __mapper_args__ = {
        "polymorphic_identity": "accounting",
    }

class Manager(User):
    __mapper_args__ = {
        "polymorphic_identity": "hr",
    }


class Team(BaseModel):
    __tablename__ = "teams"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user: Mapped[Engineer|Accounting] = relationship(back_populates="team", init=False)
# or
#  user: Mapped["Engineer"] = relationship(back_populates="team", init=False)
# or
#  user: Mapped["User"] = relationship(back_populates="team", init=False)


t1 = Team(name="team 1", id=1)
t2 = Team(name="team 2", id=2)
e = Engineer(name="Paul", id=1)
e.team = t1
a = Accounting(name="Jessica", id=2)
a.team = t2
m = Manager(name="Leto", id=3)

engine = create_engine("sqlite:///:memory:", echo=False)
BaseModel.metadata.create_all(engine)
Session = sessionmaker(engine)

print(e, a)

s = Session()
s.add_all([t1, t2, e, a, m])
s.commit()

print(s.query(User).all())
```

This model passes declaration, and runs just fine. At runtime I can add ".team" foreign keys for either instance, but when I'm saving the model I get

```
sqlalchemy.orm.exc.FlushError: Attempting to flush an item of type <class '__main__.Accounting'> as a member of collection "team.user". Expected an object of type <class '__main__.Engineer'> or a polymorphic subclass of this type.
```

Which is correct to be thrown. My relation is this after all:

```
    user: Mapped["Engineer"] = relationship(back_populates="team", init=False)
```

The docs state, I should be able to use the base class, or at least the class containing the foreign key. That is not possible in this case.

```
    user: Mapped["User"] = relationship(back_populates="team", init=False)
```

does not pass declaration:

```
sqlalchemy.exc.InvalidRequestError: Mapper 'Mapper[User(users)]' has no property 'team'
```

This is also legit. The key is in sub-class only, after all. 

Next I've tried this; but this does not appear to be the way either (it passes declaration in this sample excerpt, but for some reason not in the actual code I've stumbled upon this) - but anyway, at runtime it just seems to take whatever class is listed first in the Union.  

```
    user: Mapped[Engineer|Accounting] = relationship(back_populates="team", init=False)
```

Giving, on commit again: 

```
AssertionError: Attribute 'user' on class '<class '__main__.Team'>' doesn't handle objects of type '<class '__main__.Accounting'>'
``` 

At this point I have two questions:
* is maybe my design pattern wrong, and if so what would be the correct one?
* if not, how do I correctly declare and annotate this case?

I understand that actual subclasses of polymorphic data models are not supported, i.e. something like

```
class HasTeams(User, TeamMixIn):
    ...

class Engineer(HasTeams):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

class Accounting(HasTeams):
    __mapper_args__ = {
        "polymorphic_identity": "accounting",
    }

```

which would then allow me to use ``Mapped[HasTeams]`` on the left side of the declaration of the relationship. 

Though, after all, this _seems_ still to work. However, in my real world model, this breaks as soon as I try to have more than one base class to inherit (I do have actual permutations of 6 sub-types from two mixins in my real model, that each have different foreign keys in different combinations).

Is there a way to declare the left/annotated side of a relation _by feature_ and not by actual Type? I suppose, this a prime example for [typing protocols](https://docs.python.org/3/library/typing.html#protocols), where Python seems to push for upstream. But maybe I'm just overcomplicating the model for no reason.

Now, of course I could just move the foreign keys up into the base User model, but that somehow defeats the purpose of using classe inheritance. 

Just for the record, the example above is simplified, in my real world model, I do have an AssociationProxy in between for a m:n relationship with extra data. Getting it to work there, may possibly be the follow-up problem.

**Comment by zzzeek** [comment]
I'd have to spend some time with this, do you have a model that does what you want in 1.4 right now?    it seems like you are asking the models here to do things that were not possible in 1.4.   2.0 does not actually add any new model capabilities, it just provides new syntaxes.

**Comment by daemonkeeper** [comment]
I just tried to use 1.4 declarative syntax but I still end up with the same result (``AssertionError: Attribute 'user' on class '<class '__main__.Team'>' doesn't handle objects of type '<class '__main__.Engineer'>``"). I have no idea how I could even express anything like this in imperative, but I guess that syntax won't have much of a future anyway. 

So, yes you may be right that 1.4 was not able to do something alike either. This brings me back to my other question: if this is not supported and it appears, I am the first person ever to have this problem: what would be a better way to model this? Moving all foreign keys up sure works, but at that point I don't need to use inherited classes either. 

I'd argue my abstraction is not completely unreasonable though. I could imagine lots of use cases where you'd use class inheritance for more specific instances of a broader general type, that have foreign keys into third party tables by their identity.

**Reply by zzzeek** [reply]
whats the goal, "which would then allow me to use Mapped[HasTeams] " ?  Can you just illustrate in a single file the complete scenario you envision without any conditionals?  thanks

if the request is, "use a non-mapped mixin as a relationship target" or something like that, that would not be possible.  relationship has to always have a mapped class as its target.

**Reply by daemonkeeper** [reply]
I need "_HasTeams_" as an alias I can use in places, where I can describe the other end of a mapping by a feature, instead of an actual relation (i.e.; class). In my case the feature would be defined as "has a foreign key relationship with the mapped class".

I've very much given you an example already, Here it is, with small modifications:

```python
from sqlalchemy import ForeignKey, create_engine
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Mapped, mapped_column, relationship, declared_attr, sessionmaker
from typing import Union

class BaseModel(MappedAsDataclass, DeclarativeBase, kw_only=True):
    ...

class TeamMixIn:
    team_id: Mapped[int|None] = mapped_column(ForeignKey("teams.id"), init=False, use_existing_column=True, repr=False)

    @declared_attr.cascading
    def team(_cls) -> Mapped["Team"]:
        return relationship(back_populates="user", init=False, repr=True)


class User(BaseModel):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user_type: Mapped[str] = mapped_column(init=False, repr=False)

    __mapper_args__ = {
        "polymorphic_identity": "user",
        "polymorphic_on": "user_type",
    }

class HasTeamsFK(User, TeamMixIn):
    """
    NOT a mapped class. Must not be instantiated. This is just a generic alias for implementing sub-classes
    """
    ...

class Engineer(HasTeamsFK):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

class Accounting(HasTeamsFK):
    __mapper_args__ = {
        "polymorphic_identity": "accounting",
    }

class Manager(User):
    __mapper_args__ = {
        "polymorphic_identity": "hr",
    }


class Team(BaseModel):
    __tablename__ = "teams"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user: Mapped[HasTeamsFK] = relationship(back_populates="team", init=False)


t1 = Team(name="team 1", id=1)
t2 = Team(name="team 2", id=2)
e = Engineer(name="Paul", id=1)
e.team = t1
a = Accounting(name="Jessica", id=2)
a.team = t2
m = Manager(name="Leto", id=3)

engine = create_engine("sqlite:///:memory:", echo=False)
BaseModel.metadata.create_all(engine)
Session = sessionmaker(engine)

print(e, a)

s = Session()
s.add_all([t1, t2, e, a, m])
s.commit()

print(s.query(User).all())
```

This does, surprisingly, seem to work, albeit with a scary warning (which is legit though):

```
scratch_1.py:27: SAWarning: Mapper Mapper[HasTeamsFK(users)] does not indicate a polymorphic_identity, yet is part of an inheritance hierarchy that has a polymorphic_on column of 'users.user_type'.  Objects of this type cannot be loaded polymorphically which can lead to degraded or incorrect loading behavior in some scenarios.  Please establish a polmorphic_identity for this class, or leave it un-mapped.  To omit mapping an intermediary class when using declarative, set the '__abstract__ = True' attribute on that class.
  class HasTeamsFK(User, TeamMixIn):
Engineer(id=1, name='Paul', team=Team(id=1, name='team 1', user=[...])) Accounting(id=2, name='Jessica', team=Team(id=2, name='team 2', user=[...]))
[Engineer(id=1, name='Paul', team=Team(id=1, name='team 1', user=[...])), Accounting(id=2, name='Jessica', team=Team(id=2, name='team 2', user=[...])), Manager(id=3, name='Leto')]
```

If I make this 

```
class HasTeamsFK(User, TeamMixIn):
    __abstract__ = True
````

as suggested, I get this:

``sqlalchemy.orm.exc.UnmappedClassError: Class '__main__.HasTeamsFK' is not mapped``

Again, very much correct to state so. I do, in fact, try to to relate to this class as if it was mapped even though I would never see an instance of "HasTeamsFK" anywhere:

```
    user: Mapped[HasTeamsFK] = relationship(back_populates="team", init=False)
```

Now, I could accept this workaround for the time being, but practice shows, as soon as I start to mix-in other features like, say like this:

```
class HasTeamsFK(User, TeamMixIn):
    ...


class HasSomeotherFK(User, SomeOtherMixInSimilarToAbove):
    ...

class Engineer(HasTeamsFK, HasSomeotherFK):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }
```

I get ``sqlalchemy.exc.InvalidRequestError: Class <class '__main__.Engineer'> has multiple mapped bases: [<class '__main__.HasTeamsFK'>, <class '__main__.HasSomeotherFK'>]``.  Though I do not _really_ try to map multiple classes. 

I could possibly imagine the following ways to solve this (assuming neither of you tells me my modelling is wrong, and I should be doing it completely different anyway):

* ``__abstract__ = True`` intermediate classes must be allowed as an incoming reference in relationship() or annotated Mapped[Leftside] relations, while, at runtime ensure that the instantiated class suffices for polymorphic_identity inheritance rules
* Allow relationship / Mapped[Leftside] to specify _the actual MixIn_ as per https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html# so long, as the mapped class inherits from it 
* Allow Mapped type annotations to be [protocols](https://docs.python.org/3/library/typing.html#protocols) instead of actual types
* Allow Mapped[MappedClassA | MappedClassB] on the left side of annotations, or in relationship() arguments denoting the relation could be either of these types


On the incoming side of a mapped relation you should not really care to which actual class this relates to, so long as it is clear to which table the relation maps to, and if that mapped class has the attributes you need (in my case: "team") attribute. In my use of  ``HasTeamsFK`` this is really a purely abstract base class that guarantees two things:

* the class inherits TeamMixIn columns (and thus has the attribute the relation needs to look out for)
* it is a descendent from User / a mapping for the "users" table

And of course, this should work with more than one abstract intermediate class. 

Hope that makes clear what I am trying to achieve.

**Reply by zzzeek** [reply]
OK let me just note, I'm really pressed for time today and overall, so I read your example , but your bullets I am glossing over for now, because just at the code example I can show the issue.  if you can tell me what you expect to see, that would inform what we might do:

```py
class Team(BaseModel):
    __tablename__ = "teams"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user: Mapped[HasTeamsFK] = relationship(back_populates="team", init=False)
```

What SQL is emitted if I write this query?

```py
stmt = select(Team).join(Team.user)
```

?

**Reply by daemonkeeper** [reply]
> OK let me just note, I'm really pressed for time today and overall, so I read your example , but your bullets I am glossing over for now, because just at the code example I can show the issue. if you can tell me what you expect to see, that would inform what we might do:
> 
> ```python
> class Team(BaseModel):
>     __tablename__ = "teams"
>     id: Mapped[int] = mapped_column(primary_key=True)
>     name: Mapped[str]
>     user: Mapped[HasTeamsFK] = relationship(back_populates="team", init=False)
> ```

Not entirely sure to follow, but I'd expect:

* Manager, Engineer and Accounting records to be stored in the "users" table
* Instances of Engineer, Accounting class mappings to have a 1:1 "team" relationship to the Team/"teams" table, through the "team_id" foreign key
* Team to have a reverse relationship (above I called this "incoming relationship" before) into the users table, with instances to be either Engineer or Accounting, but never Manager
* All of the above still working by adding a third table in a different relation (e.g.; say Engineer having a 1:1 relation into Site with a similar relationship expression as above in between, but Accounting would not)
* All of the above also working in m:n relationships (with an association proxy in between to map)


> What SQL is emitted if I write this query?
> 
> ```python
> stmt = select(Team).join(Team.user)
> ```


This code 
```
print(select(Team).join(Team.user))
print(s.query(Team).all())
```

gives: 

```
SELECT teams.id, teams.name 
FROM teams JOIN users ON teams.id = users.team_id AND users.user_type IN (NULL, :user_type_1, :user_type_2)
[Team(id=1, name='team 1', user=[Engineer(id=1, name='Paul', team=...)]), Team(id=2, name='team 2', user=[Accounting(id=2, name='Jessica', team=...)])]
```

The created join statement appears alright to me. So does the session queries' result.

**Reply by zzzeek** [reply]
if `select(Team).join(Team.user)` creates a JOIN to the `users` table, then this relationship is unambiguously mapped to the `User` table.    I guess the limitation you're going for (noting again, I am not reading all the bullets here, just getting the top-level idea) is that you don't want Team.user to be a base User object, it would always be one of those subclasses.        So you would need to have a single common class between Engineer and Accounting that is mapped, not `__abstract__` , and that's the actual target of `relationship()`.  You can put whatever you want in the ``Mapping`` part though.

as far as having multiple "bases" like HasTeamsFK / HasSomeOtherFK, as soon as you want one more of those, you need to not use `__abstract__` for more than one of them, they'd have to be declarative mixins.

So if it were me, this would look like:

```py
class HasTeamsFKMixin(TeamMixIn):
    pass

class UserHasTeams(HasTeamsFKMixin, User):
    __mapper_args__ = {
        "polymorphic_identity": "_NA",
    }
    def __init__(self, *arg, **kw):
        raise NotImplementedError()


class Engineer(UserHasTeams):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }


class Accounting(UserHasTeams):
    __mapper_args__ = {
        "polymorphic_identity": "accounting",
    }


class Manager(User):
    __mapper_args__ = {
        "polymorphic_identity": "hr",
    }


class Team(BaseModel):
    __tablename__ = "teams"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user: Mapped[HasTeamsFKMixin] = relationship(UserHasTeams, back_populates="team", init=False)

```

**Reply by zzzeek** [reply]
the hypothetical change to SQLAlchemy would be some way to have single-inh mapped targets that have no polymorphic identity.  something like:

```py
class UserHasTeams(HasTeamsFKMixin, User):
    __mapper_args__ = {
        "no_poly_identity": True
    }
```

but that wouldnt really make things that much simpler.    relationship() always needs a target that is mapped to a table or selectable of some kind but it does not have to be on the `Mapped[]` annotation side.

**Reply by CaselIT** [reply]
> - Allow Mapped[MappedClassA | MappedClassB] on the left side of annotations, or in relationship() arguments denoting the relation could be either of these types

I wonder if supporting this in relationship would be an huge undertaking

**Reply by zzzeek** [reply]
it would most certainly be quite huge

**Reply by zzzeek** [reply]
i mean , as written it would be impossible.  if MappedClassA and MappedClassB are mapped to different tables, have different loader options, etc.  it's not possible.  there has to be a class that is the common base of MappedClassA / MappedClassB that is the target of the relationship().   you can put whatever you want in `Mapped` but a mapped class has to be on the right side.   We can make it "nicer" to create classes like this, but the user has to make the choice what the real-mapped target of the relationship is, we can't guess from a Union like that.

**Reply by daemonkeeper** [reply]
First off, I think we're on the same page now. 

> So you would need to have a single common class between Engineer and Accounting that is mapped, not __abstract__ , and that's the actual target of relationship(). You can put whatever you want in the Mapping part though.

This does, as I understand it, paraphrase the problem and aligns with my problem statement. 

> i mean , as written it would be impossible. if MappedClassA and MappedClassB are mapped to different tables, have different loader options, etc. it's not possible. there has to be a class that is the common base of MappedClassA / MappedClassB that is the target of the relationship(). 

While one can specify MappedClassA | CompletelyUnrelatedClassB, at runtime this should most certainly not pass building the declarative base. All you had to do is to ensure at runtime that all members of that union have a common ancestor and point to the same base relation. 
More precisely, ensure they inherit the same _features_ you need (in this case: the foreign key in the declarative base class you relate to). The actual mapping you choose can then be any member of the union, as they can be used interchangeably as far as the incoming relation is concerned, but flushing should take any instance in that union.

**Reply by daemonkeeper** [reply]
> the hypothetical change to SQLAlchemy would be some way to have single-inh mapped targets that have no polymorphic identity. something like:
> 
> ```python
> class UserHasTeams(HasTeamsFKMixin, User):
>     __mapper_args__ = {
>         "no_poly_identity": True
>     }
> ```
> 
> but that wouldnt really make things that much simpler. relationship() always needs a target that is mapped to a table or selectable of some kind but it does not have to be on the `Mapped[]` annotation side.

That also has the drawback to make the modeling quite ugly. You have HasTeamsFKMixin with column/relation mixins, and you have another class doing nothing but putting itself into the class inerhitance chain of inerhited table classes, and inherting from the Mixin. You'd want to be able to merge those two classes then.

**Reply by zzzeek** [reply]
>  All you had to do is to ensure at runtime that all members of that union have a common ancestor from the same base relation. 

I agree with the second half of that sentence but not the first half, "all you had to do", as though heuristics like this are simple, and not, as it turns out, things that produce new bugs for months or years.   Checking for "a common base" is not sufficient.  Mapped classes, including subclasses of a common base, may be mapped to different tables, have different polymorphic settings, have different event listeners associated with them, have different additional relationships and attributes with impacts on subsequent loader operations, etc.    It is also my experience that features like these which seem "simple" usually turn out to have way more edge cases than even the few I can think of at the moment.     Beyond that, relationship links to a single mapper.  without the presence of a common base class, that's pretty much a complete rewrite for relationship() with an open ended number of new complexities.   All of which would be to support a use case that is not widely requested and in fact is only being requested here.    

There's actually a ratio that I consult which is "simplicity of feature / broadness of userbase request" - a very simple feature with no implications beyond the new feature is often sufficient even if only one user requests it.  A very complex feature that requires re-thinking core architectural decisions from the ground up OTOH would require a very broad user desire, such as our work to add pep-484 typing throughout SQLAlchemy, which in fact is not as big of a change as it would be for relationship() to know how to link to more than one mapper.

Since re-architecting relationship() to refer to mulitple target mappers is a non-starter, it follows that the feature is provided by creating a common base class between the desired targets. Given that, it's most appropriate that the user supplies this base.  It can be documented, have examples, tests, syntactic helpers, fine.   Rewriting relationship() to make it "implicit", that would not be a good undertaking.

**Reply by daemonkeeper** [reply]
I can't argue with that. I am really not endorsing to rewrite the entire ``relationship`` business logic. Moreover, you are definitely in a better position to assess the actual impact of a change like this. However, as I've shown the actual _implementation_ of it is already satisfying the needs. It does already do all the heavy lifting in a seemingly correct way, as the queries and session saves show.  

It just needs a few hints in the right place, to actually accept _generalized_ intermediate types (I suppose you'd still call them "mappings") to pass the fenced and rightfully gated security checks in cases where this would not be an actually supported use case. I have honestly no idea what would be the best way to do, if you were to implement this change request, thus I made some suggestions I could think of.

I believe, I raised reasonable arguments why the design I modeled my classes for is not _entirely_ random and follows a clear design idea as in why you'd possibly want to have it modeled in such a way. If you all agree, this is a super special use case I can sure deal with that, and just won't use table class inheritance at all in this case. I felt like bringing it up nonetheless here, because you provide all the set pieces needed to make this a clear and slick abstraction, I just can't bring them together. 

I guess, the bottom line is: my exact modelling is not currently doable in SQLAlchemy, and it's not a bug either. It's a wishlist new feature. How would you like to move on from here?

**Reply by zzzeek** [reply]
> I can't argue with that. I am really not endorsing to rewrite the entire `relationship` business logic. Moreover, you are definitely in a better position to assess the actual impact of a change like this. However, as I've shown the actual _implementation_ of it is already satisfying the needs. It does already do all the heavy lifting in a seemingly correct way, as the queries and session saves show.

it does!  your working example includes the `HasTeams` class without `__abstract__ = True`, which means that it's mapped.  There's just the issue of some warnings generated about polymorphic identity not set up, which is mostly a cosmetic issue.

> 
> It just needs a few hints in the right place, to actually accept _generalized_ intermediate types (I suppose you'd still call them "mappings") to pass the fenced and rightfully gated security checks in cases where this would not be an actually supported use case. I have honestly no idea what would be the best way to do, if you were to implement this change request, thus I made some suggestions I could think of.

Yes I agree, we just want some kind of `__abstract__` setting that still allows the class to be "mapped".


> 
> I believe, I raised reasonable arguments why the design I modeled my classes for is not _entirely_ random and follows a clear design idea as in why you'd possibly want to have it modeled in such a way. If you all agree, this is a super special use case I can sure deal with that, and just won't use table class inheritance at all in this case. 

I think what you're doing is fine.  If there are more "XHasFKs" things though, those have to be pure mixins.   We also dont support a class that extends from multiple mapped classes at once (we have a little more request for this every few years, but not meeting that need/complexity ratio).

> 
> I guess, the bottom line is: my exact modelling is not currently doable in SQLAlchemy, and it's not a bug either. It's a wishlist new feature. How would you like to move on from here?

I think we just need some directive so that your intermediary "generalized" class remains "mapped", but doesnt emit a warning for polymorphic identity not included and perhaps has some kind of "dont instantiate" behavior attached to it, and that would get us 95% of the way there.  what do you think?

**Reply by zzzeek** [reply]
noting however the other limitation that you could only extend from one of these classes at a time.  if you wanted to include multiple "HasXYZFK" bases, those have to be mixins that you'd combine to one common base that's in the mapped class hierarchy.

**Comment by daemonkeeper** [comment]
Just so that I understand you: you're proposing something like:

```
class User(BaseModel):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    user_type: Mapped[str] = mapped_column(init=False, repr=False)

    __mapper_args__ = {
        "polymorphic_identity": "user",
        "polymorphic_on": "user_type",
    }

class HasSomeFK(User, TeamMixIn):
    __some_new_mapper_flag__ = True

class HasSomeOtherFK(User, AnotherFKMixIn):
    __some_new_mapper_flag__ = True

class HasBothFKs(User, AnotherFKMixIn, TeamMixIn):
    __some_new_mapper_flag__ = True


class Engineer(HasSomeFK):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

class Accounting(HasSomeOtherFK):
    __mapper_args__ = {
        "polymorphic_identity": "accounting",
    }

class Manager(HasBothFKs):
    __mapper_args__ = {
        "polymorphic_identity": "hr",
    }
```

With either of the intermediate classes being allowed to be put on the left side of an annotated statement in a ``relationship()``?

**Reply by zzzeek** [reply]
yeah that is the proposal.    Alternatively, you could put those intermediate classes on the **right** side within `relationship()`, and put anything you want on the left side.

**Reply by daemonkeeper** [reply]
The syntax isn't really beautiful I'd dare to say, but it gets the job done. Should I create a wishlist "new use case" ticket for you?

**Reply by zzzeek** [reply]
let's name the attribute here first.   I propose `__no_instantiate__ = True`

**Reply by daemonkeeper** [reply]
You could also consider using the existing ``__abstract__`` attribute, and define it as

``True``, ``full``: fully abstract class (the existing meaning of the attribute; backwards-compatible, while suggesting in the docs to not use a boolean for it in new code)
``virtual``, or ``polymorphic``: for the use-case above

I have no share in the actual choice, but I'd argue that people may find it confusing when to use ``__abstract__`` vs ``__no_instantiate__ `` if both do something seemingly identical. You may even be getting into loopholes if I was definining both at the same time, while they should be mutually exclusive, I guess.

**Reply by zzzeek** [reply]
yes, probably use `__abstract__`, it's always a little cringe with variables that are `True/ False / <string> / <string>` but we obviously have tons of these already

**Reply by daemonkeeper** [reply]
Now that you started to natively support dataclasses, and you support Enums natively as a data type, you could also start adding them to your API. In my code, I am a big advocate for something like 

```python
from enum import Enum, auto

class POLYMORPHIC_MAPPING_BEHAVIOR(Enum):
     NONE = auto()
     FULLY_ABSTRACT = auto()
     MAPPABLE = auto()
```

and in user's code

```python 
class MyMapping(Base):
     __abstract__ = POLYMORPHIC_MAPPING_BEHAVIOR.FULLY_ABSTRACT

```

But then, I am not normally in a position that I need to support 16 years of happy users out there, and changes like this are not really backwards-compatible. On the upside they're more future-proof as you can easily add and scope new possible values in.

**Reply by zzzeek** [reply]
we use a lot of enums but we haven't made the users use them.    I would not want to start making some arguments here and there into enums without a plan to make the whole library go there, which I dont know if that's going to happen anytime soon.  consistency in API usage is important so I try to stay away from using multiple conventions in the library.

**Reply by zzzeek** [reply]
also we use typing.Literal[] to type these, which at the typing level is just as good as an enum; your IDE will highlight a wrong value anyway.   so that's another reason a full transfer to enum is  likely unnecessary.

**Reply by daemonkeeper** [reply]
> we use a lot of enums but we haven't made the users use them. I would not want to start making some arguments here and there into enums without a plan to make the whole library go there, which I dont know if that's going to happen anytime soon. consistency in API usage is important so I try to stay away from using multiple conventions in the library.

Oh yeah, I fully agree. A change like this should be consistent and across all of the API. And most definitely coming along with other changes that are backwards incompatible and inevitable. Thanks for the Literal[] hint by the way, I didn't know about that one yet. 

Anyway, I was just having some tech convo. You are of course the person to judge much better than I, what is the most consistent way to integrate this.

**Reply by zzzeek** [reply]
yeah it's going to be the usual True/False/"here"/"there" thing...   it's fine, just added "auto" for mapper.eager_defaults the other day

**Reply by daemonkeeper** [reply]
Happy new year to all of you! How do we proceed from here on? Would you create an issue/feature request about it, or shall I? I think we discussed everything at this point.

**Reply by CaselIT** [reply]
If you can open a feature request we can better track it

**Reply by daemonkeeper** [reply]
Thanks! I've created https://github.com/sqlalchemy/sqlalchemy/issues/9060. Hopefully it summarizes our discussion well enough.
